### ISSUE TITLE
Goto Tag with Remark Color

### ISSUE THREAD

**Issue by Zaidy036** [ScintillaOrg/lexilla#148]
Minor bug in Notepad++ configured for Batch

When adding a comment on a goto tag command line the entire line has the goto tag color.

Sample:
:DropboxC & ::     Limit is 5 Start Attempts
The complete line above will have the goto tag color AND will function correctly as the goto tag.

**Comment by mpheath** [comment]
It is known as a label. Type `goto /?` or `call /?` in CMD for help description.

![batch_label](https://user-images.githubusercontent.com/58158242/224258558-75e12717-538a-41cd-b229-f2d6a286ffed.png)

<details>

<summary>As text:</summary>

```
echo text & rem
echo text & ::comment
:label & echo text
:label & ::comment
```

</details>

The first 2 lines are interpreted as runtime code.

The last 2 lines are labels. Labels are accessed with `call` or `goto`, which causes the interpreter to stop execution and to literally read the file for the label and start execution from the following line. Any text after the label name on the same line is ignored by the interpreter and since is not runtime code makes no sense to style as runtime code. Example: line 3 will not echo the word `text`. If line 3 is not valid runtime code, then neither is line 4. There is nothing to fix.

**Comment by Zaidy036** [comment]
Sorry that I used incorrect terminology.

For Batch Default I have configured:
LABELS as Bold Red on Yellow background
COMMENTS as Bold Dark Green on White background.

A LABEL with a COMMENT on the same line
           :Label & :: Comment
is entirely Bold Red on Yellow background. The presence of the comment does not affect the batch operation but is just not as I have it configured.

If it can be fixed, great, but it is not a major problem.

**Comment by nyamatongwe** [comment]
There is no comment in your sample. There is a label `:DropboxC` and some other text ` & :: Limit is 5 Start Attempts` that has no effect.

The line could be written without a `::` and the effect would be the same:
~~~
:DropboxC Limit is 5 Start Attempts
~~~

**Comment by mpheath** [comment]
I consider the runtime code styling idea as bad for a *label*, though perhaps an alternative to text after `::` as automatic comment style. The patch is just to see what it looks like.

<details>

<summary>View patch</summary>

``` diff
--- C:/Users/Michael/AppData/Local/Temp/s3fg.1	Sat Mar 11 15:13:11 2023
+++ D:/Projects/Scite_Sc1/make_scite/lexilla/lexers/LexBatch.cxx	Sat Mar 11 15:08:22 2023
@@ -159,7 +159,23 @@
 					styler.ColourTo(endPos, SCE_BAT_COMMENT);
 				} else {
 					// Colorize Real Label
-					styler.ColourTo(endPos, SCE_BAT_LABEL);
+					bool changed = false;
+					while (offset < lengthLine) {
+						offset++;
+						if (lineBuffer[offset] == ':') {
+							if (offset + 1 < lengthLine) {
+								if (lineBuffer[offset + 1] == ':') {
+									styler.ColourTo(startLine + offset - 1, SCE_BAT_LABEL);
+									styler.ColourTo(endPos, SCE_BAT_COMMENT);
+									changed = true;
+									break;
+								}
+							}
+						}
+					}
+					if (!changed) {
+						styler.ColourTo(endPos, SCE_BAT_LABEL);
+					}
 				}
 				stopLineProcessing=true;
 			// Check for Drive Change (Drive Change is internal command) - return if found

```

Test
```
:DropboxC
:DropboxC :: Limit is 5 Start Attempts
:DropboxC:: Limit is 5 Start Attempts
:DropboxC
```

Styled
```
{3}:DropboxC
:DropboxC {1}:: Limit is 5 Start Attempts
{3}:DropboxC{1}:: Limit is 5 Start Attempts
{3}:DropboxC
```

</details>

[Issue148.zip](https://github.com/ScintillaOrg/lexilla/files/10947887/Issue148.zip)

Unfortunately, the *EOLFilled* style cuts off once the *comment* style starts so may look wierd.

![batch_label_2](https://user-images.githubusercontent.com/58158242/224466675-c86c4be8-5f62-4e25-8fec-606016dc5949.png)

That is the unmodified styling of SciTE.

I am not sure if it is worth it or would even use it as not a fan of inline comments. This may upset users who already use `::` in *label* lines and do not expect the styling to change. It may also encourage bad code formatting as users may see this to exploit and possibly want to push into other lexer projects. I am not sure yet of any buggy effects to doing this.

Is this idea worth it?

**Comment by Zaidy036** [comment]
nyamatongwe is correct. 

:LABEL functions up to the first space and anything after has no effect.

That is different from other Batch commands and in itself is a contradiction in general philosophy of command form in Batch.

I would call that a bug in the :LABEL itself.

**Comment by nyamatongwe** [comment]
There could be an extra 'failure' style (`SCE_BAT_EXTRANEOUS`?) for any text after a label to show that it won't be interpreted. Those that like adding descriptions after labels could choose an appearance similar to comments and others could show it as a problem with a red background or similar.

**Comment by zufuliu** [comment]
in Notepad2, I choose to highlight text after label as comment as it has not effect.
![image](https://user-images.githubusercontent.com/2289926/225617443-fd79abd6-a1ee-44d4-9657-6b580ec0a82e.png)

**Comment by mpheath** [comment]
> There could be an extra 'failure' style (`SCE_BAT_EXTRANEOUS`?) for any text after a label to show that it won't be interpreted. Those that like adding descriptions after labels could choose an appearance similar to comments and others could show it as a problem with a red background or similar.

Seems like a this or that and not both scenerio. I'm not sure how the lexer code can know how to handle this or that. Probably an option added later could switch the behaviour.

I warmed up to this idea once I tested it.

<details>

<summary>Click to view image</summary>

![label_extra](https://user-images.githubusercontent.com/58158242/225854765-619bc9a3-049c-40e8-a2d3-3a967bd57755.gif)

2 frames to view.
All yellow text as default. Yellow text and green text as custom.

Testing with `IsASpaceOrTab()` to set `SCE_BAT_EXTRANEOUS` so `::` is irrelevent with the new style.
<hr>
</details>

`SCE_BAT_LABEL` + `SCE_BAT_COMMENT` looks undesirable to me. Notepad2 gets away with it as the background of the label style is almost completely transparent so the contrast with the styles is not too different, though still subtly noticable. `SCE_BAT_LABEL` + `SCE_BAT_EXTRANEOUS` is better as `SCE_BAT_EXTRANEOUS` can be styled independently to blend in with `SCE_BAT_LABEL`. The feature can visually separate by style the label name from the following text instead of looking like one combined sentence of words.

Extraneous is not a word I see often. I guess it is unrelated until the point of use which here is text following the label name.

I regularly add parameter names after label names and [read](https://github.com/mpheath/make-scite-collection/blob/main/base/batch/extension.lua) them for autocomplete/calltips, though not comments with labels accessed by `goto` as this issue is first mention of the idea.

**Comment by nyamatongwe** [comment]
With `SCE_BAT_EXTRANEOUS` I was trying to be slightly disapproving since text after a label has no defined role. `SCE_BAT_ERROR` or  `SCE_BAT_UNDEFINED` are a bit strong. `SCE_BAT_AFTER_LABEL` may be more neutral.

**Comment by nyamatongwe** [comment]
Treating after-label as comment:
![BAT_AFTER_LABEL_comment](https://user-images.githubusercontent.com/225818/226219476-c6f56320-f4ed-49be-b074-9f85d1b6dae9.png)

Treating after-label as problem:
![BAT_AFTER_LABEL](https://user-images.githubusercontent.com/225818/226219479-d44890c9-700f-4ea9-8c67-ff41cf6ea674.png)

A patch to experiment with (the #define will go in a header):
~~~patch
@@ -159,7 +159,15 @@ void ColouriseBatchDoc(
 					styler.ColourTo(endPos, SCE_BAT_COMMENT);
 				} else {
 					// Colorize Real Label
-					styler.ColourTo(endPos, SCE_BAT_LABEL);
+					#define SCE_BAT_AFTER_LABEL 8
+					const char *startLabelName = lineBuffer + offset + 1;
+					const char *endLabel = strpbrk(startLabelName, " \t:");
+					if (endLabel) {
+						styler.ColourTo(startLine + offset + endLabel - startLabelName, SCE_BAT_LABEL);
+						styler.ColourTo(endPos, SCE_BAT_AFTER_LABEL);	// New style
+					} else {
+						styler.ColourTo(endPos, SCE_BAT_LABEL);
+					}
 				}
 				stopLineProcessing=true;
 			// Check for Drive Change (Drive Change is internal command) - return if found
~~~
Other characters could go in the `strpbrk` but couldn't find a definition of characters allowed in a label.

**Comment by mpheath** [comment]
>Other characters could go in the [`strpbrk`](https://cplusplus.com/reference/cstring/strpbrk/) but couldn't find a definition of characters allowed in a label.

Seemly valid label name terminators:

``` py
['\t', '\n', '\r', ' ', '&', '+', ':', '<', '>', '|']
```

Would consider testing `\t &+:<>|`

A label could be `:label:`, `:label>` ... with the last character coloured as the new style. I suppose someone could do `:label:comment` if considered more desirable without any spaces.

<details>

<summary>Click for Python script to get terminators</summary>

``` py
import subprocess

valid = []
chars = list(range(9, 14)) + list(range(32, 127))

for i in chars:
    # 0-9
    if i >= 48 and i <= 57:
        continue

    # A-Z
    if i >= 65 and i <= 90:
        continue

    # a-z
    if i >= 97 and i <= 122:
        continue

    with open('test_label.cmd', 'w') as w:
        w.write('@goto label\n:label{}\n'.format(chr(i)))

    with subprocess.Popen('test_label.cmd',
                          stderr=subprocess.DEVNULL,
                          shell=True) as p:
        p.wait()

        if not p.returncode:
            valid.append(chr(i))

print('valid:', valid)
```

</details>

**Comment by mpheath** [comment]
May have a problem?

``` batch
goto label
echo will not see
: label
echo found label
```

`label` is found, though the styling is not correct.

![batch_label_error](https://user-images.githubusercontent.com/58158242/226511991-6127beea-7d05-4f7f-acea-3d0048a8fe00.png)

`goto`/`call` seems to cause the parser to remove `: ` to find `label`. Sorry, I didn't know this behaviour before now. Tested with space ` ` and tab `\t` before the label name which are stripped.

**Comment by nyamatongwe** [comment]
857cc50a32d8b248b09766363d09598185e58198 ignores space between `:` and label name.

**Comment by nyamatongwe** [comment]
Included in release 5.2.5.
